docs/book/src/tasks/experimental-features/cluster-classes.md

```diff
@@ -0,0 +1,497 @@
+# Cluster Class
+
+The ClusterClass feature introduces a new way to create clusters which reduces boilerplate required and allows flexible and powerful customization of CAPI-created clusters.
+ClusterClass is implemented on top of pre-existing CAPI interfaces and will 
+
+This tutorial covers using Cluster API and Cluster Class to create a Kubernetes cluster and how to perform a one-touch Kubernetes update procedure using ClusterClass features.
+
+<aside class="note warning">
+
+<h1>Note</h1>
+
+This guide covers creating a ClusterClass cluster with the CAPD provider.
+</aside>
+
+## Installation
+
+### Common Prerequisites
+This guide is based on having the same pre-requisites as the CAPI [quick start guide].
+That is:
+- Install and setup [kubectl]  and [clusterctl]
+- Install [Kind] and [Docker]
+
+ClusterClass is currently behind a feature gate that needs to be enabled. To do so run:
+
+```bash
+export  EXP_CLUSTER_RESOURCE_SET=true
+export CLUSTER_TOPOLOGY=true
+```
+This ensures that Cluster Topology and Cluster Resource Set are enabled when we initialize our CAPI Providers.
+*Note*: Cluster Resource Set is not required for ClusterClass but is used in this tutorial to set up networking.
+
+A script to set up a Kind cluster pre-configured for CAPD can be found in the hack folder. To set up the cluster run:
+```bash
+./hack/kind-install-for-capd.sh
+clusterctl init --infrastructure docker
+````
+
+Alternatively the quickstart guide for setting up a CAPD enabled cluster can be followed.
+
+#### Creating the ClusterClass and templates
+Now that we have a CAPI Management Cluster with CAPD and Cluster Topology enabled we can create the ClusterClass and the templates needed to stamp clusters.
+
+The below contains six yaml objects which are used to define the cluster:
+
+1) ClusterClass
+1) DockerMachineTemplate
+1) DockerClusterTemplate
+1) DockerMachineTemplate for ControlPlane
+1) KubeadmControlPlaneTemplate
+1) KubeadmConfigTemplate
+
+The ClusterClass - first in the below yaml - contains references to the templates needed to build a full cluster using the shape defined in the ClusterClass.
+```yaml
```

Comment:
Just an idea. Should we move this file and the next to: ./test/infrastructure/docker/examples and then just provide a kubectl apply command? Would make the quickstart way more compact.

Reply:
@fabriziopandini what do you think? I know the clusterctl generate flow is on the way, but it would make the guide more compact and less error prone I think.

Reply:
I'm ok with both solutions. Up to you

Reply:
Wondering what the best solution is here - I've left the clusterclass templates and the clusteryaml in the guide so that we're explicit about the objects that users are working with, but I've moved them under the test/infrastructure package so that they can be easily created with kubectl apply - any thoughts on the current form? I think showing the yamls in the tutorial has some value. Do we have any mechanism for embedding yamls in the book?

Reply:
I'm fine with the current version. Not sure if there's a way to wrap it in something which can be expanded/collapsed on demand (I couldn't find an example in our book)

Reply:
@killianmuldoon I think you can try `details` if you want, it's used in the kubebuilder book:
* https://book.kubebuilder.io/cronjob-tutorial/basic-project.html
* https://raw.githubusercontent.com/kubernetes-sigs/kubebuilder/master/docs/book/src/cronjob-tutorial/basic-project.md

Looks like a good compromise to me.

Reply:
Ah now I got what you mean with embedding. Afaik you can use things like `{{#include ./testdata/project/PROJECT}}` (also shown in the kubebuilder example I linked)

Reply:
I've done this and moved the yamls to be inside the book folder rather than in examples. This might be more appropriate for now - especially if those were going to be removed if clusterctl generate replaces raw yamls here. WDYT?

Reply:
I think you're right.